File: src/main/java/de/silencio/activecraftcore/guicreator/GuiItem.java
package de.silencio.activecraftcore.guicreator;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import org.bukkit.Material;
import org.bukkit.Sound;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.ArrayList;
import java.util.List;

@Getter
@ToString
@EqualsAndHashCode(callSuper = false)
public class GuiItem extends ItemStack {

    private Sound clickSound = Sound.UI_BUTTON_CLICK;
    private boolean movable;
    private final List<ClickListener> clickListenerList = new ArrayList<>();

    public GuiItem(Material material) {
        super(material);
    }

    public GuiItem(Material material, int stackSize) {
        super(material, stackSize);
    }

    public GuiItem(ItemStack itemStack) {
        super(itemStack != null ? itemStack.getType() : Material.AIR);
        if (itemStack == null) return;
        this.setItemMeta(itemStack.getItemMeta());
        this.setAmount(itemStack.getAmount());
        for (ItemFlag itemFlag : itemStack.getItemFlags()) {
            this.addItemFlags(itemFlag);
        }
    }

    public GuiItem setDisplayName(String displayName) {
        ItemMeta itemMeta = this.getItemMeta();
        itemMeta.setDisplayName(displayName);
        this.setItemMeta(itemMeta);
        return this;
    }

    public GuiItem setLore(String... lore) {
        List<String> stringList = new ArrayList<>(List.of(lore));
        ItemMeta itemMeta = this.getItemMeta();
        itemMeta.setLore(stringList);
        this.setItemMeta(itemMeta);
        return this;
    }

    public GuiItem setLore(int index, String lore) {
        ItemMeta itemMeta = this.getItemMeta();
        List<String> loreList = itemMeta.getLore() != null ? itemMeta.getLore() : new ArrayList<>();
        loreList.set(Math.min(loreList.size(), index), lore);
        setLore(loreList);
        return this;
    }

    public GuiItem addLore(String... lore) {
        ItemMeta itemMeta = this.getItemMeta();
        List<String> loreList = itemMeta.getLore() != null ? itemMeta.getLore() : new ArrayList<>();
        loreList.addAll(List.of(lore));
        setLore(loreList);
        return this;
    }

    public GuiItem removeLore(int i) {
        ItemMeta itemMeta = this.getItemMeta();
        List<String> loreList = new ArrayList<>();
        if (itemMeta.getLore() != null) loreList.addAll(itemMeta.getLore());
        loreList.remove(i);
        setLore(loreList);
        return this;
    }

    public GuiItem removeLore(int start, int end) {
        for (int i = start; i < end; i++) removeLore(i);
        return this;
    }

    public String getDisplayName() {
        return this.getItemMeta().getDisplayName();
    }

    public List<String> getLore() {
        return this.getItemMeta().getLore();
    }

    public GuiItem setClickSound(Sound clickSound) {
        this.clickSound = clickSound;
        return this;
    }

    public GuiItem setMovable(boolean movable) {
        this.movable = movable;
        return this;
    }

    public ItemStack toItemStack() {
        return this;
    }

    public GuiItem setGlint(boolean glint) {
        if (glint) {
            addUnsafeEnchantment(Enchantment.DURABILITY, 1);
            addItemFlags(ItemFlag.HIDE_ENCHANTS);
        } else if (getEnchantmentLevel(Enchantment.DURABILITY) == 1) {
            removeEnchantment(Enchantment.DURABILITY);
        }
        return this;
    }

    public GuiItem addClickListener(ClickListener clickListener) {
        this.clickListenerList.add(clickListener);
        return this;
    }
}

File: src/main/java/org/activecraft/activecraftcore/guicreator/GuiItem.kt
package org.activecraft.activecraftcore.guicreator

import org.bukkit.Material
import org.bukkit.Sound
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack

open class GuiItem : ItemStack {
    var clickSound: Sound? = Sound.UI_BUTTON_CLICK
        protected set
    var movable = false
        protected set
    val clickListenerList: MutableList<(GuiClickEvent) -> Unit> = mutableListOf()

    @JvmOverloads
    constructor(material: Material, stackSize: Int = 1) : super(material, stackSize)
    constructor(itemStack: ItemStack) : super(itemStack.type) {
        this.itemMeta = itemStack.itemMeta
        amount = itemStack.amount
        for (itemFlag in itemStack.itemFlags) {
            addItemFlags(itemFlag!!)
        }
    }

    fun setDisplayName(displayName: String): GuiItem {
        val itemMeta = this.itemMeta
        itemMeta.setDisplayName(displayName)
        this.itemMeta = itemMeta
        return this
    }

    fun setLore(vararg lore: String): GuiItem {
        val stringList: List<String> = listOf(*lore)
        val itemMeta = this.itemMeta
        itemMeta.lore = stringList
        this.itemMeta = itemMeta
        return this
    }

    fun setLore(index: Int, lore: String): GuiItem {
        val itemMeta = this.itemMeta
        val loreList = if (itemMeta.lore != null) itemMeta.lore else ArrayList()
        loreList!![loreList.size.coerceAtMost(index)] = lore
        setLore(loreList)
        return this
    }

    fun addLore(vararg lore: String): GuiItem {
        val itemMeta = this.itemMeta
        val loreList = if (itemMeta.lore != null) itemMeta.lore else ArrayList()
        loreList!!.addAll(listOf(*lore))
        setLore(loreList)
        return this
    }

    fun removeLore(i: Int): GuiItem {
        val itemMeta = this.itemMeta
        val loreList: MutableList<String> = ArrayList()
        if (itemMeta.lore != null) loreList.addAll(itemMeta.lore!!)
        loreList.removeAt(i)
        lore = loreList
        return this
    }

    fun removeLore(start: Int, end: Int): GuiItem {
        for (i in start until end) removeLore(i)
        return this
    }

    val displayName: String
        get() = this.itemMeta.displayName

    override fun getLore() = itemMeta.lore

    fun setClickSound(clickSound: Sound?): GuiItem {
        this.clickSound = clickSound
        return this
    }

    fun setMovable(movable: Boolean): GuiItem {
        this.movable = movable
        return this
    }

    fun toItemStack() = this as ItemStack

    fun setGlint(glint: Boolean): GuiItem {
        if (glint) {
            addUnsafeEnchantment(Enchantment.DURABILITY, 1)
            addItemFlags(ItemFlag.HIDE_ENCHANTS)
        } else if (getEnchantmentLevel(Enchantment.DURABILITY) == 1) {
            removeEnchantment(Enchantment.DURABILITY)
        }
        return this
    }

    fun addClickListener(clickListener: (GuiClickEvent) -> Unit): GuiItem {
        clickListenerList.add(clickListener)
        return this
    }
}